src/btree/bt_read.c

```diff
@@ -8,21 +8,400 @@
+		if (__wt_txn_visible_all(session, txnid))
+			continue;
+
+		/*
+		 * Skip past the counter (it's only needed to ensure records are
+		 * read in their original, listed order), leaving p referencing
+		 * the first byte of the key.
+		 */
+		p = (uint8_t *)p + sizeof(uint32_t);
+
+		/* Allocate the WT_UPDATE structure. */
+		WT_ERR(cursor->get_value(cursor, &txnid, &upd_size, vlas));
+		WT_ERR(__wt_update_alloc(session,
+		    (upd_size == WT_UPDATE_DELETED_VALUE) ? NULL : vlas,
+		    &upd, &incr));
+		total_incr += incr;
+		upd->txnid = txnid;
+
+		switch (page->type) {
+		case WT_PAGE_COL_FIX:
+		case WT_PAGE_COL_VAR:
+			memcpy(&recno, p, sizeof(uint64_t));
+			if (current_recno == recno)
+				break;
+
+			if (first_upd != NULL) {
+				WT_WITH_DHANDLE(session, read_dhandle,
+				    ret = __col_update(session,
+				    current_recno, ref, &cbt, first_upd));
+				WT_ERR(ret);
+
+				first_upd = NULL;
+			}
+			current_recno = recno;
+			break;
+		case WT_PAGE_ROW_LEAF:
+			memcpy(&key_len, p, sizeof(uint32_t));
+			p = (uint8_t *)p + sizeof(uint32_t);
+			if (current_key->size == key_len && memcmp(
+			    current_key->data, p, current_key->size) == 0)
+				break;
+
+			if (first_upd != NULL) {
+				WT_WITH_DHANDLE(session, read_dhandle,
+				    ret = __row_update(session,
+				    current_key, ref, &cbt, first_upd));
+				WT_ERR(ret);
+
+				first_upd = NULL;
+			}
+			WT_ERR(__wt_buf_set(session, current_key, p, key_len));
+			break;
+		WT_ILLEGAL_VALUE_ERR(session);
+		}
+
+		/* Append the latest update to the list. */
+		if (first_upd == NULL)
+			first_upd = last_upd = upd;
+		else {
+			last_upd->next = upd;
+			last_upd = upd;
+		}
+		upd = NULL;
+	}
+	WT_ERR_NOTFOUND_OK(ret);
+
+	/* Insert the last set of updates, if any. */
+	if (first_upd != NULL)
+		switch (page->type) {
+		case WT_PAGE_COL_FIX:
+		case WT_PAGE_COL_VAR:
+			WT_WITH_DHANDLE(session, read_dhandle,
+			    ret = __col_update(session,
+			    current_recno, ref, &cbt, first_upd));
+			WT_ERR(ret);
+
+			first_upd = NULL;
+			break;
+		case WT_PAGE_ROW_LEAF:
+			WT_WITH_DHANDLE(session, read_dhandle,
+			    ret = __row_update(session,
+			    current_key, ref, &cbt, first_upd));
+			WT_ERR(ret);
+
+			first_upd = NULL;
+			break;
+		WT_ILLEGAL_VALUE_ERR(session);
+		}
+
+	/* Discard the cursor. */
+	WT_ERR(__wt_las_cursor_close(session, &cursor, saved_flags));
+
+	if (total_incr != 0) {
+		__wt_cache_page_inmem_incr(session, page, total_incr);
+
+		/*
+		 * We modified the page above, which will have set the first
+		 * dirty transaction to the last transaction currently running.
+		 * However, the updates we installed may be older than that.
+		 * Set the first dirty transaction to an impossibly old value
+		 * so this page is never skipped in a checkpoint.
+		 *
+		 * KEITH: is this correct? We don't care about checkpoints, we
+		 * care about older readers in the system.
+		 */
+		page->modify->first_dirty_txn = WT_TXN_FIRST;
+	}
```

Comment:
`first_dirty_txn` is only relevant for genuinely dirty pages (for checkpoints).  Restored pages with updates in LAS should start clean, right?